File: atm/python/atm.py
class Account:
    next_id = 0

    def __init__(self, pin: str, init_balance: int):
        self.id = "A" + str(Account.next_id)
        Account.next_id += 1
        self.pin = pin
        self.curr_balance = init_balance


class Card:
    next_id = 0

    def __init__(self, account_id: str):
        self.id = "C" + str(Card.next_id)
        Card.next_id += 1
        self.account_id = account_id


class Bank:
    def __init__(self):
        self.accounts: dict[str, Account] = {}
        self.cards: dict[str, Card] = {}
    
    def add_new_account(self, pin: str, init_balance: int) -> str:
        account = Account(pin, init_balance)
        self.accounts[account.id] = account
        return account.id
        
    def add_new_card(self, account_id: str) -> str | None:
        if account_id not in self.accounts:
            return None
        
        card = Card(account_id)
        self.cards[card.id] = card
        return card.id
    
    def is_valid_card(self, card_id):
        return card_id in self.cards
    
    def authenticate(self, card_id: str, pin: str) -> bool:
        card = self.cards[card_id]
        account = self.accounts[card.account_id]
        return account.pin == pin
    
    def get_balance(self, card_id: str) -> int:
        card = self.cards[card_id]
        account = self.accounts[card.account_id]
        return account.curr_balance
    
    def deposit(self, card_id: str, amount: int) -> int | None:
        card = self.cards[card_id]
        account = self.accounts[card.account_id]
        account.curr_balance += amount
        return account.curr_balance
    
    def withdraw(self, card_id: str, amount: int) -> int | None:
        card = self.cards[card_id]
        account = self.accounts[card.account_id]
        
        if account.curr_balance < amount:
            return None
        
        account.curr_balance -= amount
        return account.curr_balance
    

class ATM:
    def __init__(self, init_cash_balance: int, bank: Bank):
        self.bank = bank
        self.curr_cash_balance = init_cash_balance
        self.session = False
        self.session_card_id = None
        self.session_authenticated = False

    def insert_card(self, card_id: str) -> bool:
        if self.session:
            return False
        
        if not self.bank.is_valid_card(card_id):
            return False

        self.session = True
        self.session_card_id = card_id
        self.session_authenticated = False
        return True
    
    def enter_pin(self, pin: str) -> bool:
        if not self.session:
            return False
        
        self.session_authenticated = self.bank.authenticate(self.session_card_id, pin)
        return self.session_authenticated
    
    def check_balance(self) -> int | None:
        if not self.session:
            return None
        
        if not self.session_authenticated:
            return None
        
        return self.bank.get_balance(self.session_card_id)

    def deposit_money(self, amount: int) -> int | None:
        if amount <= 0:
            return None
        
        if not self.session:
            return None
        
        if not self.session_authenticated:
            return None
        
        balance = self.bank.deposit(self.session_card_id, amount)
        self.curr_cash_balance += amount
        return balance

    def withdraw_money(self, amount: int) -> int | None:
        if amount <= 0:
            return None
        
        if not self.session:
            return None
        
        if not self.session_authenticated:
            return None
        
        if self.curr_cash_balance < amount:
            return None
        
        balance = self.bank.withdraw(self.session_card_id, amount)
        
        if balance is None:
            return None
        
        self.curr_cash_balance -= amount
        return balance
    
    def eject_card(self) -> None:
        self.session = False
        self.session_card_id = None
        self.session_authenticated = False

def main():
    bank = Bank()
    account_id = bank.add_new_account(pin="0000", init_balance=50)
    card_id = bank.add_new_card(account_id=account_id)
    atm = ATM(init_cash_balance=100, bank=bank)

    assert atm.insert_card(card_id)
    assert atm.enter_pin("0000")
    new_balance = atm.withdraw_money(10)
    assert new_balance == 40
    atm.eject_card()

if __name__ == "__main__":
    main()

File: atm/cpp/atm.cpp
#include <string>
#include <unordered_map>
#include <optional>
#include <memory>   // for smart pointers (std::unique_ptr)

class Account {
private:
    inline static int next_id = 0;

public:
    std::string id;
    std::string pin;
    int curr_balance;

    Account(std::string pin, int init_balance)
        : id("A" + std::to_string(next_id)),
          pin(pin),
          curr_balance(init_balance)
    {
        ++next_id;
    }
};

class Card {
private:
    inline static int next_id = 0;

public:
    std::string id;
    std::string account_id;

    Card(std::string account_id)
        : id("C" + std::to_string(next_id)),
          account_id(account_id)
    {
        ++next_id;
    }
};

class Bank {
public:
    std::unordered_map<std::string, std::unique_ptr<Account>> accounts;
    std::unordered_map<std::string, std::unique_ptr<Card>> cards;

    Bank() {}

    std::string add_new_account(std::string pin, int init_balance) {
        auto account = std::make_unique<Account>(pin, init_balance);
        std::string account_id = account->id;

        accounts.emplace(account_id, std::move(account));

        return account_id;
    }

    std::optional<std::string> add_new_card(std::string account_id) {
        if (accounts.find(account_id) == accounts.end()) {
            return std::nullopt;
        }

        auto card = std::make_unique<Card>(account_id);
        std::string card_id = card->id;

        cards.emplace(card_id, std::move(card));

        return card_id;
    }

    bool is_valid_card(std::string card_id) {
        return cards.find(card_id) != cards.end();
    }

    bool authenticate(std::string card_id, std::string pin) {
        if (!is_valid_card(card_id)) {
            return false;
        }

        Card* card = cards.at(card_id).get();
        Account* account = accounts.at(card->account_id).get();

        return account->pin == pin;
    }

    int get_balance(std::string card_id) {
        Card* card = cards.at(card_id).get();
        Account* account = accounts.at(card->account_id).get();

        return account->curr_balance;
    }

    int deposit(std::string card_id, int amount) {
        Card* card = cards.at(card_id).get();
        Account* account = accounts.at(card->account_id).get();

        account->curr_balance += amount;

        return account->curr_balance;
    }

    std::optional<int> withdraw(std::string card_id, int amount) {
        Card* card = cards.at(card_id).get();
        Account* account = accounts.at(card->account_id).get();

        if (account->curr_balance < amount) {
            return std::nullopt;
        }

        account->curr_balance -= amount;

        return account->curr_balance;
    }
};

class ATM {
public:
    int curr_cash_balance;
    Bank& bank;
    bool session;
    std::string session_card_id;
    bool session_authenticated;

    ATM(int init_cash_balance, Bank& bank)
        : curr_cash_balance(init_cash_balance),
          bank(bank),
          session(false),
          session_card_id(""),
          session_authenticated(false)
    {}

    bool insert_card(std::string card_id) {
        if (session) {
            return false;
        }

        if (!bank.is_valid_card(card_id)) {
            return false;
        }

        session = true;
        session_card_id = card_id;
        session_authenticated = false;

        return true;
    }

    bool enter_pin(std::string pin) {
        if (!session) {
            return false;
        }

        session_authenticated = bank.authenticate(session_card_id, pin);

        return session_authenticated;
    }

    std::optional<int> check_balance() {
        if (!session) {
            return std::nullopt;
        }

        if (!session_authenticated) {
            return std::nullopt;
        }

        return bank.get_balance(session_card_id);
    }

    std::optional<int> deposit_money(int amount) {
        if (amount <= 0) {
            return std::nullopt;
        }

        if (!session) {
            return std::nullopt;
        }

        if (!session_authenticated) {
            return std::nullopt;
        }

        int balance = bank.deposit(session_card_id, amount);
        curr_cash_balance += amount;

        return balance;
    }

    std::optional<int> withdraw_money(int amount) {
        if (amount <= 0) {
            return std::nullopt;
        }

        if (!session) {
            return std::nullopt;
        }

        if (!session_authenticated) {
            return std::nullopt;
        }

        if (curr_cash_balance < amount) {
            return std::nullopt;
        }

        std::optional<int> balance = bank.withdraw(session_card_id, amount);

        if (!balance.has_value()) {
            return std::nullopt;
        }

        curr_cash_balance -= amount;

        return balance;
    }

    void eject_card() {
        session = false;
        session_card_id = "";
        session_authenticated = false;
    }
};